src/core/lib/transport/metadata.h

```diff
@@ -132,17 +134,135 @@ void* grpc_mdelem_get_user_data(grpc_mdelem md, void (*if_destroy_func)(void*));
+  grpc_slice key;
+  grpc_slice value;
+
+  /* private only data */
+  gpr_atm refcnt;
+
+  UserData user_data;
+
+  ~allocated_metadata();
+} allocated_metadata;
+
 #ifndef NDEBUG
 #define GRPC_MDELEM_REF(s) grpc_mdelem_ref((s), __FILE__, __LINE__)
-#define GRPC_MDELEM_UNREF(s) grpc_mdelem_unref((s), __FILE__, __LINE__)
-grpc_mdelem grpc_mdelem_ref(grpc_mdelem md, const char* file, int line);
-void grpc_mdelem_unref(grpc_mdelem md, const char* file, int line);
+inline grpc_mdelem grpc_mdelem_ref(grpc_mdelem gmd, const char* file,
+                                   int line) {
 #else
 #define GRPC_MDELEM_REF(s) grpc_mdelem_ref((s))
+inline grpc_mdelem grpc_mdelem_ref(grpc_mdelem gmd) {
+#endif
+  switch (GRPC_MDELEM_STORAGE(gmd)) {
+    case GRPC_MDELEM_STORAGE_EXTERNAL:
+    case GRPC_MDELEM_STORAGE_STATIC:
+      break;
+    case GRPC_MDELEM_STORAGE_INTERNED: {
+      interned_metadata* md =
+          reinterpret_cast<interned_metadata*> GRPC_MDELEM_DATA(gmd);
+#ifndef NDEBUG
+      grpc_mdelem_trace_ref(md, md->key, md->value, &md->refcnt, file, line);
+#endif
+      /* we can assume the ref count is >= 1 as the application is calling
+         this function - meaning that no adjustment to mdtab_free is necessary,
+         simplifying the logic here to be just an atomic increment */
+      /* use C assert to have this removed in opt builds */
+      GPR_DEBUG_ASSERT(gpr_atm_no_barrier_load(&md->refcnt) >= 1);
+      gpr_atm_no_barrier_fetch_add(&md->refcnt, 1);
+      break;
+    }
+    case GRPC_MDELEM_STORAGE_ALLOCATED: {
+      allocated_metadata* md =
+          reinterpret_cast<allocated_metadata*> GRPC_MDELEM_DATA(gmd);
+#ifndef NDEBUG
+      grpc_mdelem_trace_ref(md, md->key, md->value, &md->refcnt, file, line);
+#endif
+      /* we can assume the ref count is >= 1 as the application is calling
+         this function - meaning that no adjustment to mdtab_free is necessary,
+         simplifying the logic here to be just an atomic increment */
+      /* use C assert to have this removed in opt builds */
+      gpr_atm_no_barrier_fetch_add(&md->refcnt, 1);
+      break;
+    }
+  }
+  return gmd;
+}
+
+void grpc_note_disposed_interned_metadata(uint32_t hash);
+#ifndef NDEBUG
+#define GRPC_MDELEM_UNREF(s) grpc_mdelem_unref((s), __FILE__, __LINE__)
+inline void grpc_mdelem_unref(grpc_mdelem gmd, const char* file, int line) {
+#else
 #define GRPC_MDELEM_UNREF(s) grpc_mdelem_unref((s))
-grpc_mdelem grpc_mdelem_ref(grpc_mdelem md);
-void grpc_mdelem_unref(grpc_mdelem md);
+inline void grpc_mdelem_unref(grpc_mdelem gmd) {
+#endif
+  switch (GRPC_MDELEM_STORAGE(gmd)) {
+    case GRPC_MDELEM_STORAGE_EXTERNAL:
+    case GRPC_MDELEM_STORAGE_STATIC:
+      break;
+    case GRPC_MDELEM_STORAGE_INTERNED: {
+      interned_metadata* md =
+          reinterpret_cast<interned_metadata*> GRPC_MDELEM_DATA(gmd);
+#ifndef NDEBUG
+      grpc_mdelem_trace_unref(md, md->key, md->value, &md->refcnt, file, line);
+#endif
+      uint32_t hash = md->hash;
```

Comment:
I'd move this load into the branch. That's the only place it's used right?

if you do that I think we can inline unref too by using the same method for both types:

```
if (1 == prev_refcount) {
   grpc_mdelem_destroy(md);
}
```

Reply:
There will be an aliasing issue, but that's easy to work around:
https://godbolt.org/z/1uISrP

```
void grpc_mdelem_destroy(grpc_mdelem gmd);
void grpc_mdelem_unref(grpc_mdelem gmd) {
  bool destroy = false;
  switch (GRPC_MDELEM_STORAGE(gmd)) {
    case GRPC_MDELEM_STORAGE_EXTERNAL:
    case GRPC_MDELEM_STORAGE_STATIC:
      break;
    case GRPC_MDELEM_STORAGE_INTERNED: {
      interned_metadata* md =
          reinterpret_cast<interned_metadata*> GRPC_MDELEM_DATA(gmd);
      
      const intptr_t prev_refcount = md->refcnt.fetch_sub(1, std::memory_order_acq_rel);
      destroy = 1 == prev_refcount;
      break;
    }
    case GRPC_MDELEM_STORAGE_ALLOCATED: {
      allocated_metadata* md =
          reinterpret_cast<allocated_metadata*> GRPC_MDELEM_DATA(gmd);
      const intptr_t prev_refcount = md->refcnt.fetch_sub(1, std::memory_order_acq_rel);
      destroy = 1 == prev_refcount;
      break;
    }
  }
  if (destroy) { grpc_mdelem_destroy(gmd); }
}
```

Reply:
The load cannot be moved into the branch, since the act of unref() could cause another thread to free() md before we read md->hash.

Consequently, unable to make this specific change. As a compromise, I've split the unref() into the inlined (static+external) and uninlined (interned+allocated) segments. Unref() is now 8 instructions when inlined